src/v/jl/jls_writer/jls_writer.cc

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "jl/jls_writer/jls_writer.h"
+
+#include "base/likely.h"
+#include "base/vassert.h"
+#include "bytes/iobuf.h"
+#include "bytes/iostream.h"
+#include "jl/jls/jls.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "model/record_batch_types.h"
+#include "serde/rw/rw.h"
+
+#include <seastar/core/future.hh>
+
+#include <__utility/move.h>
+
+#include <stdexcept>
+#include <stdint.h>
+#include <utility>
+
+namespace {
+void write_ntp(iobuf& buf, const model::ntp& ntp) {
+    serde::write(buf, ntp.ns);
+    serde::write(buf, ntp.tp.topic);
+    serde::write(buf, ntp.tp.partition);
+}
+} // namespace
+
+namespace jl::jls_writer {
+
+ss::future<> writer::write(model::ntp ntp, model::record_batch&& batch) {
+    vassert(
+      batch.header().type == model::record_batch_type::raft_data,
+      "Only raft data expected in the jls writer");
+
+    if (unlikely(_closed)) {
+        throw std::runtime_error("writer is closed");
+    }
+
+    if (unlikely(!_did_write_header)) {
+        co_await write_header();
+    }
+
+    if (ntp != _last_ntp || _chunk_buffer.empty()) {
+        if (unlikely(ntp < _last_ntp)) {
+            throw std::runtime_error(
+              "writer api invariant violation: ntps are out of order");
+        }
+
+        maybe_flush_batch_buffer();
+
+        write_ntp(_chunk_buffer, ntp);
+
+        _last_ntp = ntp;
+        _last_offset = model::offset{};
+    }
+
+    if (unlikely(_last_offset > batch.base_offset())) {
+        throw std::runtime_error(
+          "writer api invariant violation: offsets are out of order");
+    }
+
+    serde::write(_batch_buffer, std::move(batch));
```

Comment:
let's use same data representation that we're using for disk and cloud objects now.
the serialization code is in the `storage/record_batch_utils.h` the function is called `batch_header_to_disk_iobuf`.